### PR TITLE
feat(import-target): Add resolution error reason

### LINKED ISSUE #182: Bug: `import-target` mutes resolution errors
Currently in [`lib/util/import-target.js:273`](https://github.com/eslint-community/eslint-plugin-n/blob/master/lib/util/import-target.js#L273) we skip all errors as if it errors the package is not found.
However, we need to be able to display some of the errors as not all not found errors are equal!

---

An example:

My file:
```js
// index.js
import 'missing';
import 'misconfigured';
```

```json
{
  "name": "demo",
  "type": "module",
  "dependencies": {
    "misconfigured": "*"
  }
}
```

When resolving `missing` the error is:
> `Can't resolve 'soup' in '/home/scagood/issue-demo'`

When resolving `misconfigured` the error is:
> `Default condition should be last one`

In `node_modules/misconfigured/package.json` the following mistake was made:
```json
{
  "name": "misconfigured",
  "exports": {
    ".": {
      "default": "./build/index.js",
      "types": "./src/index.ts"
    }
  },
  ...
}
```

In order to resolve this, the `exports` should become:
```json
{
  "name": "misconfigured",
  "exports": {
    ".": {
      "types": "./src/index.ts",
      "default": "./build/index.js"
    }
  },
  ...
}
```

---

This is because `enhanced-resolve` explicitly throws an error here:
https://github.com/webpack/enhanced-resolve/blob/main/lib/util/entrypoints.js#L475

I am not 100% sure why... but it is an error that we need to be able to display!

### PR DESCRIPTION
**Pull request by scagood**
closes #182

### CODE REVIEW

File: lib/util/import-target.js [hunk 1]
```diff
@@ -239,6 +245,24 @@ module.exports = class ImportTarget {
...
+     */
+    handleResolutionError(baseDir, error) {
+        if (error instanceof Error === false) {
+            throw error
+        }
+
+        // This is the usual error
+        if (error.message === `Can't resolve '${this.name}' in '${baseDir}'`) {
+            return
+        }
```
Comment (aladdin-add) [inline lib/util/import-target.js:261]: I'm a bit worried that this check is error-prone: the error message may change in a later release.

Comment (scagood) [reply lib/util/import-target.js:261]: This is fair, I wonder if the error message should just be piped straight through :thinking:

Comment (aladdin-add) [reply lib/util/import-target.js:261]: 👍 from me!

**Comment by scagood** [comment]
I was also thinking something along the lines of:
```js
exports.messages = {
    notFound: '{{resolveError}}',
}
```

**Comment by aladdin-add** [comment]
SGTM! 👍

**Comment by scagood** [comment]
mmm, these failures are interesting, as on a mac the file did resolve as that os is case insensitive 🤔

image

This error is getting triggered:
```js
// This should never happen... this is just a fallback for typescript
target.resolveError ??= `"${target.name}" is not found`
```

**Comment by aladdin-add** [comment]
can we just assert the `messageId`, not `message`?

**Comment by scagood** [comment]
There was no error:
https://github.com/scagood/eslint-plugin-n/blob/muted-resolution-errors/lib/util/check-existence.js#L20-L21

What is tripping the error is:
https://github.com/scagood/eslint-plugin-n/blob/muted-resolution-errors/lib/util/check-existence.js#L59
https://github.com/scagood/eslint-plugin-n/blob/muted-resolution-errors/lib/util/check-existence.js#L81

**Comment by aladdin-add** [comment]
we can just assert the `messageId`, not the full message. wdyt?

**Comment by scagood** [comment]
The issue is that `./A.js` will resolve sometimes, OS dependent :eyes:

**Comment by aladdin-add** [comment]
ops, I misunderstood. Is there an option to config it?

**Comment by aladdin-add** [comment]
image

But when I look at the execution results, it is returning different error messages under some os. So I suggest to only assert the `messageId`. am I missing something?

**Comment by scagood** [comment]
The issue is that on your machine, `./A` does actually resolve in require

On my linux machine (which is case sensitive) it does not, so the error is correct:
image

**Comment by scagood** [comment]
I have just stripped out everything except for the enhanced-resolve checks (https://github.com/eslint-community/eslint-plugin-n/pull/231/commits/f9e454d1a24aeb168ca786ed7f41ad0e80778a8a)

(Clicking the image should take you to the tests from which the screen shot is taken)

On linux (that is case sensitive), everything is fine:
[image](https://github.com/eslint-community/eslint-plugin-n/actions/runs/8718524566/job/23915958885?pr=231)

On mac (that is case insensitive), the `./A.js` is resolvable:
[image](https://github.com/eslint-community/eslint-plugin-n/actions/runs/8718524566/job/23915960819?pr=231)

---

This means that on linux, its correct that `.tests/fixtures/no-missing/A.js` does not get loaded:

image

And its correct on mac that `.tests/fixtures/no-missing/A.js` does get loaded:

image

**Comment by aladdin-add** [comment]
we can ask the `enhanced-resolve` to support a case-sensitive option.

Up until that point, it seems that we can only ignore these test cases under certain os. The rule tester does support `only` property, but not `skip`, we will need to implement it in our side. wdyt?

**Pull request closed without merging**